torch/_guards.py

```diff
@@ -558,6 +559,12 @@ def is_nn_module(self) -> bool:
         return self.guard_source().is_nn_module()
 
 
+# Subclasses can be found in torch/_dynamo/source.py
```

Comment:
Explain a bit more, so people adding new sources know whether or not to inherit from this?

One question especially, is if I have a Source that contains multiple sources (as is the case with SuperSource), which one applies? I think this is why I wasn't a big fan of your proposed invariant here the last time we talked about it, to me Source is an AST (tree), there's no reason for it to have linked list structure.

Reply:
Sources are trees.

Every source with a base should have just a single base, with the exception of SuperSource.

SuperSource is the only one - and it does not feel worth it to do `base -> [Source]` just for that - I'd rather carve out an exception for it.

Reply:
Added a comment.